engine/src/main/java/org/terasology/config/flexible/SettingImpl.java

```diff
@@ -0,0 +1,201 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.config.flexible;
+
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.config.flexible.validators.SettingValueValidator;
+import org.terasology.engine.SimpleUri;
+
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.text.MessageFormat;
+import java.util.Set;
+
+/**
+ * {@inheritDoc}
+ *
+ * @param <T> The type of the value this {@link SettingImpl} contains.
+ */
+public class SettingImpl<T> implements Setting<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SettingImpl.class);
+
+    private final SimpleUri id;
+    private final String warningFormatString;
+
+    private final T defaultValue;
+
+    private T value;
+
+    private String humanReadableName;
+
+    private String description;
+    private SettingValueValidator<T> validator;
+    private Set<PropertyChangeListener> subscribers;
+
+    /**
+     * Creates a new {@link SettingImpl} with the given id and default value but no validator.
+     *
+     * @param id           the id of the setting.
+     * @param defaultValue the default value of the setting.
+     */
+    public SettingImpl(SimpleUri id, T defaultValue) {
+        this(id, defaultValue, null);
+    }
+
+    /**
+     * Creates a new {@link SettingImpl} with the given id, default value and validator.
+     *
+     * @param id           the id of the setting.
+     * @param defaultValue the default value of the setting.
+     * @param validator    the validator to be used to validate values.
+     */
+    public SettingImpl(SimpleUri id, T defaultValue, SettingValueValidator<T> validator) {
+        this.id = id;
+        this.warningFormatString = MessageFormat.format("Setting {0}: '{'0}'", this.id);
+
+        this.validator = validator;
+
+        if (!validate(defaultValue))
+            throw new IllegalArgumentException("The default value must be a valid value. " +
+                    "Check the logs for more information.");
+
+        this.defaultValue = defaultValue;
+        this.value = this.defaultValue;
+
+        this.subscribers = null;
+    }
+
+    private void dispatchChangedEvent(PropertyChangeEvent event) {
+        for (PropertyChangeListener subscriber : subscribers) {
+            subscriber.propertyChange(event);
+        }
+    }
+
+    private boolean validate(T value) {
+        return validator == null || validator.validate(value);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public boolean subscribe(PropertyChangeListener listener) {
+        if (subscribers == null) {
+            subscribers = Sets.newHashSet();
+        }
+
+        if (listener == null) {
+            LOGGER.warn(MessageFormat.format(this.warningFormatString,
```

Comment:
I was thinking of passing this format string as a base format string to the SettingValueValidator so that the id of the corresponding string can be printed out in the logs too.